pandas/core/groupby/grouper.py

```diff
@@ -577,8 +565,20 @@ def __repr__(self) -> str:
     def __iter__(self):
         return iter(self.indices)
 
-    _codes: np.ndarray | None = None
-    _group_index: Index | None = None
+    @cache_readonly
+    def _ilevel(self) -> int | None:
```

Comment:
why is this needed?

Reply:
i find it nicer as a property than re-calculating it in places where it is used